From ce2958d0d5374e9ae6c0d8dad63599dffb32702d Mon Sep 17 00:00:00 2001
From: Robert Lai <c_rlai@codeaurora.org>
Date: Thu, 10 Sep 2015 17:16:36 -0700
Subject: [PATCH] sched: optimize select_best_cpu() with same power cost CPU
 mask

At present, the function select_best_cpu() has to continue to iterate the
search CPUs for the possibility of finding a more power efficient CPU even
if the best CPU based on cstate already found.  We can avoid this
unnecessary search if it is known ahead of time that CPUs in a cluster
have the same power efficiency.

Introduce kernel parameter 'sched_same_power_cost_cpus' to inform the
CPU mask that has an identical power costs and optimize the 'for' loop of
select_best_cpu() by ruling out CPUs that cannot be more power efficient
than cstate based best CPU with the same power cost CPU mask.

CRs-fixed: 849655
Change-Id: I9d2e7f560282a914fd0373aac6714353b9d1e36c
Signed-off-by: Joonwoo Park <joonwoop@codeaurora.org>
---
 kernel/sched/core.c  | 20 ++++++++++++++++++++
 kernel/sched/fair.c  | 31 +++++++++++++++++++------------
 kernel/sched/sched.h |  3 +++
 3 files changed, 42 insertions(+), 12 deletions(-)

diff --git a/kernel/sched/core.c b/kernel/sched/core.c
index a0f31da..6ea6cb4 100644
--- a/kernel/sched/core.c
+++ b/kernel/sched/core.c
@@ -1083,6 +1083,26 @@ static int __init set_sched_enable_power_aware(char *str)
 
 early_param("sched_enable_power_aware", set_sched_enable_power_aware);
 
+bool have_sched_same_pwr_cost_cpus;
+cpumask_var_t sched_same_pwr_cost_cpus;
+static int __init set_sched_same_power_cost_cpus(char *str)
+{
+	char buf[64];
+
+	alloc_bootmem_cpumask_var(&sched_same_pwr_cost_cpus);
+	if (cpulist_parse(str, sched_same_pwr_cost_cpus) < 0) {
+		pr_warn("sched: Incorrect sched_same_power_cost_cpus cpumask\n");
+		return -EINVAL;
+	}
+
+	cpulist_scnprintf(buf, sizeof(buf), sched_same_pwr_cost_cpus);
+	if (!cpumask_empty(sched_same_pwr_cost_cpus))
+		have_sched_same_pwr_cost_cpus = true;
+	return 0;
+}
+
+early_param("sched_same_power_cost_cpus", set_sched_same_power_cost_cpus);
+
 static inline int got_boost_kick(void)
 {
 	int cpu = smp_processor_id();
diff --git a/kernel/sched/fair.c b/kernel/sched/fair.c
index dfce98e..c6b7116 100644
--- a/kernel/sched/fair.c
+++ b/kernel/sched/fair.c
@@ -2034,7 +2034,7 @@ static inline int wake_to_idle(struct task_struct *p)
 static int select_best_cpu(struct task_struct *p, int target, int reason,
 			   int sync)
 {
-	int i, j, prev_cpu, best_cpu = -1;
+	int i, j, best_cpu = -1;
 	int fallback_idle_cpu = -1, min_cstate_cpu = -1;
 	int cpu_cost, min_cost = INT_MAX;
 	int min_idle_cost = INT_MAX, min_busy_cost = INT_MAX;
@@ -2071,8 +2071,12 @@ static int select_best_cpu(struct task_struct *p, int target, int reason,
 	}
 
 	trq = task_rq(p);
+	i = task_cpu(p);
 	cpumask_and(&search_cpus, tsk_cpus_allowed(p), cpu_online_mask);
-	for_each_cpu(i, &search_cpus) {
+	if (unlikely(!cpumask_test_cpu(i, &search_cpus)))
+		i = cpumask_first(&search_cpus);
+
+	do {
 		struct rq *rq = cpu_rq(i);
 
 		trace_sched_cpu_load(cpu_rq(i), idle_cpu(i),
@@ -2089,12 +2093,12 @@ static int select_best_cpu(struct task_struct *p, int target, int reason,
 			continue;
 		}
 
+		cpumask_clear_cpu(i, &search_cpus);
+
 		tload =  scale_load_to_cpu(task_load(p), i);
 		if (skip_cpu(trq, rq, i, tload, reason))
 			continue;
 
-		prev_cpu = (i == task_cpu(p));
-
 		/*
 		 * The least-loaded mostly-idle CPU where the task
 		 * won't fit is our fallback if we can't find a CPU
@@ -2172,15 +2176,19 @@ static int select_best_cpu(struct task_struct *p, int target, int reason,
 				min_idle_cost = cpu_cost;
 				min_cstate = cstate;
 				min_cstate_cpu = i;
-				continue;
-			}
 
-			if (cpu_cost < min_idle_cost ||
-			    (prev_cpu && cpu_cost == min_idle_cost)) {
+				if (prefer_idle &&
+				    have_sched_same_pwr_cost_cpus &&
+				    min_cstate == 0 &&
+				    cpumask_test_cpu(i,
+					sched_same_pwr_cost_cpus))
+					cpumask_andnot(&search_cpus,
+						&search_cpus,
+						sched_same_pwr_cost_cpus);
+			} else if (cpu_cost < min_idle_cost) {
 				min_idle_cost = cpu_cost;
 				min_cstate_cpu = i;
 			}
-
 			continue;
 		}
 
@@ -2203,12 +2211,11 @@ static int select_best_cpu(struct task_struct *p, int target, int reason,
 		 * This is rare but when it does happen opt for the
 		 * more power efficient CPU option.
 		 */
-		if (cpu_cost < min_busy_cost ||
-		    (prev_cpu && cpu_cost == min_busy_cost)) {
+		if (cpu_cost < min_busy_cost) {
 			min_busy_cost = cpu_cost;
 			best_cpu = i;
 		}
-	}
+	} while ((i = cpumask_first(&search_cpus)) < nr_cpu_ids);
 
 	/*
 	 * Don't need to check !sched_cpu_high_irqload(best_cpu) because
diff --git a/kernel/sched/sched.h b/kernel/sched/sched.h
index 0c6b2ae..3190ee4 100644
--- a/kernel/sched/sched.h
+++ b/kernel/sched/sched.h
@@ -262,6 +262,9 @@ struct hmp_power_cost_table {
 	struct hmp_power_cost *map;
 };
 
+extern bool have_sched_same_pwr_cost_cpus;
+extern cpumask_var_t sched_same_pwr_cost_cpus;
+
 #endif
 
 /* CFS-related fields in a runqueue */
-- 
1.9.1

